pupil-api/src/services/redis.service.ts

```diff
@@ -0,0 +1,176 @@
+import Redis, { RedisOptions } from 'ioredis'
+import config from '../config'
+import { Logger } from './log.service'
+
+export interface IRedisService {
+  /**
+   * @description retrieve an item from the cache, under the given key
+   * @param {string} key the unique string key of the redis entry to fetch
+   * @throws when the data type of the retrieved value is unsupported
+   * @returns {Promise<string | null>} an awaitable promise containing the item if it exists, or undefined if it does not
+   */
+  get (key: string): Promise<any | null>
+  /**
+   * @description insert or ovewrite an item in the cache, which lives for a specific duration
+   * @param {string} key the unique string key of the redis entry to persist
+   * @param {object | string} value the item to persist in redis cache
+   * @param {number} ttl how long to store the item in seconds
+   * @throws when the incoming item datatype is not supported and when the setex redis operation fails
+   * @returns {Promise<void} an awaitable promise
+   */
+  setex (key: string, value: string | object, ttl: number): Promise<void>
+  /**
+   * @description drop a series of items from the cache
+   * @param {Array<string>} keys an array of keys to invalidate
+   * @returns {Promise<void>}
+   */
+  drop (keys: Array<string>): Promise<void>
+  /**
+   * @description cleans up the underlying redis client implementation
+   * @returns void
+   */
+  quit (): Promise<string>
+}
+
+export class RedisService implements IRedisService {
+
+  private redis: Redis.Redis
+  private logger: Logger
+
+  constructor () {
+    const options: RedisOptions = {
+      port: Number(config.Redis.Port),
+      host: config.Redis.Host,
+      password: config.Redis.Key
+    }
+    if (config.Redis.useTLS) {
+      options.tls = {
+        host: config.Redis.Host
+      }
+    }
+    this.redis = new Redis(options)
+   /* this.redis
+     .on('connect', () => {
+      this.logger.info('redis:connect')
+    })
+    .on('ready', () => {
+      this.logger.info('redis:ready')
+    })
+    .on('error', (e) => {
+      this.logger.info('redis:ready', e)
+    })
+    .on('close', () => {
+      this.logger.info('redis:close')
+    })
+    .on('reconnecting', () => {
+      this.logger.info('redis:reconnecting')
+    })
+    .on('end', () => {
+      this.logger.info('redis:end')
+    }) */
+    this.logger = new Logger()
+  }
+
+  async get (key: string): Promise<any | null> {
+    try {
+      const cacheEntry = await this.redis.get(key)
+      if (cacheEntry === null) return Promise.resolve(null)
+      const cacheItem: RedisCacheItem = JSON.parse(cacheEntry)
```

Comment:
Currently, the admin app is not setting the payload to be a typed RedisCacheItem.  It would be nice if it did though.

Reply:
yes, it would.  🤔

Reply:
separate ticket created for this